carbon-mss/components/org.wso2.carbon.mss/src/main/java/org/wso2/carbon/mss/internal/deployer/MSSJarProcessor.java

```diff
@@ -0,0 +1,128 @@
+     */
+    public MSSJarProcessor setArtifact(File file) {
+        this.file = file;
+        return this;
+    }
+
+    /**
+     * Initialize a list of objects from the jar for the specified
+     * classes in the jar's manifest file under 'microservices' key
+     */
+    public MSSJarProcessor process() {
+        String jarPath = file.getAbsolutePath();
+        String[] mssClassNames = readMSSManifestEntry(jarPath);
+        URLClassLoader classLoader;
+        try {
+            //Parent class loader is required to provide classes that are outside of the jar
+            //A Privileged Block is not required to initialise class loader since this code is internal
+            classLoader = new URLClassLoader(new URL[]{file.toURI().toURL()}, this.getClass().getClassLoader());
+        } catch (MalformedURLException e) {
+            log.error("Path to jar is invalid");
+            return this;
+        }
+        for (String className : mssClassNames) {
+            try {
+                Class classToLoad = classLoader.loadClass(className);
+                resourceInstances.add(classToLoad.newInstance());
+                log.info("initialized class: " + className);
+            } catch (ClassNotFoundException e) {
+                log.error("Class: " + className + " not found");
+            } catch (InstantiationException e1) {
+                log.error("Failed to initialize class: " + className);
+            } catch (IllegalAccessException e1) {
+                log.error("Failed to access class: " + className);
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Extracts the comma separated list of fully qualified class
+     * names of the 'microservices' key of the jar's maifest file
+     *
+     * @param jarPath absolute path to the jar file
+     * @return String array of fully qualified class names
+     */
+    private String[] readMSSManifestEntry(String jarPath) {
+        JarFile jarFile = null;
+        try {
+            jarFile = new JarFile(jarPath);
+            Manifest manifest = jarFile.getManifest();
+            if (manifest == null) {
+                log.error("Error retrieving manifest: " + jarPath);
+                return new String[0];
+            }
+            Attributes mainAttributes = manifest.getMainAttributes();
+            String mssEntry = mainAttributes.getValue(MICROSERVICES_MANIFEST_KEY);
+            if (mssEntry == null) {
+                log.error("\'microservices\' manifest entry not found: " + jarPath);
+                return new String[0];
+            }
+            return mssEntry.split("\\s*,\\s*");
+        } catch (IOException e) {
+            log.error("Error retrieving manifest: " + jarPath);
+            return new String[0];
+        } finally {
+            if (jarFile != null) {
+                try {
+                    jarFile.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
```

Comment:
printStacktrace can't be located anywhere in the code now so it seems this
has been fixed.

On Sun, Jan 10, 2016 at 2:34 PM, Afkham Azeez <azeez@wso2.com> wrote:

> Wonder how that e.printstacktrace escaped checkstyles?
>
> On Sun, Jan 10, 2016 at 10:11 AM, Mohanadarshan Vivekanandalingam <
> notifications@github.com> wrote:
>
>> In
>> carbon-mss/components/org.wso2.carbon.mss/src/main/java/org/wso2/carbon/mss/internal/deployer/MSSJarProcessor.java
>> <https://github.com/wso2/product-mss/pull/3#discussion_r49271568>:
>>
>> > +            String mssEntry = mainAttributes.getValue(MICROSERVICES_MANIFEST_KEY);
>> > +            if (mssEntry == null) {
>> > +                log.error("\'microservices\' manifest entry not found: " + jarPath);
>> > +                return new String[0];
>> > +            }
>> > +            return mssEntry.split("\\s*,\\s*");
>> > +        } catch (IOException e) {
>> > +            log.error("Error retrieving manifest: " + jarPath);
>> > +            return new String[0];
>> > +        } finally {
>> > +            if (jarFile != null) {
>> > +                try {
>> > +                    jarFile.close();
>> > +                } catch (IOException e) {
>> > +                    e.printStackTrace();
>> > +                }
>>
>> This is not the correct way to log the exception.. This statement will
>> not log the exception in log file.. Please fix this..
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/wso2/product-mss/pull/3/files#r49271568>.
>>
>
>
>
> --
> *Afkham Azeez*
> Director of Architecture; WSO2, Inc.; http://wso2.com
> Member; Apache Software Foundation; http://www.apache.org/
> * <http://www.apache.org/>*
> *email: **azeez@wso2.com* <azeez@wso2.com>
> * cell: +94 77 3320919 <%2B94%2077%203320919>blog: *
> *http://blog.afkham.org* <http://blog.afkham.org>
> *twitter: **http://twitter.com/afkham_azeez*
> <http://twitter.com/afkham_azeez>
> *linked-in: **http://lk.linkedin.com/in/afkhamazeez
> <http://lk.linkedin.com/in/afkhamazeez>*
>
> *Lean . Enterprise . Middleware*
>



-- 
*Afkham Azeez*
Director of Architecture; WSO2, Inc.; http://wso2.com
Member; Apache Software Foundation; http://www.apache.org/
* <http://www.apache.org/>*
*email: **azeez@wso2.com* <azeez@wso2.com>
* cell: +94 77 3320919blog: **http://blog.afkham.org*
<http://blog.afkham.org>
*twitter: **http://twitter.com/afkham_azeez*
<http://twitter.com/afkham_azeez>
*linked-in: **http://lk.linkedin.com/in/afkhamazeez
<http://lk.linkedin.com/in/afkhamazeez>*

*Lean . Enterprise . Middleware*